ir/state.cpp

```diff
@@ -343,12 +342,48 @@ bool State::isUndef(const expr &e, const Value *used_by) const {
   return undef_vars.count(e) != 0;
 }
 
+smt::expr State::mkFreshUndef(const smt::expr &e_ty, bool do_register) {
+  expr newv;
+
+  auto same_type = [](const expr &e1, const expr &e2) {
```

Comment:
you should implement a method in expr to check type. There it's much cheaper (just check if sort() are ==)